Integration test failing

Integration test failed in the job that run when #173 was merged to main. We may not be cleaning up test resources properly.

From the [logs](https://pipelines.actions.githubusercontent.com/serviceHosts/df6c6f84-92d2-41c9-ac99-f5287d989bd0/_apis/pipelines/1/runs/578/signedlogcontent/29?urlExpires=2023-05-10T03%3A22%3A14.3180904Z&urlSigningMethod=HMACV1&urlSignature=z6Aqjmaq1UqisMcxoqjg0GpjJymbfkAqQoQ%2BfUQf9Ig%3D):

```
2023-05-09T21:12:56.6642706Z ....................2023-05-09 21:12:56,663 |    33 |                 gunicorn.error |    ERROR | Connection in use: ('0.0.0.0', 8060)
2023-05-09T21:12:56.6656639Z 2023-05-09 21:12:56,665 |    33 |                 gunicorn.error |    ERROR | Retrying in 1 second.
2023-05-09T21:12:57.6723942Z /root/.local/lib/python3.11/site-packages/gunicorn/sock.py:-1: ResourceWarning: unclosed <socket.socket fd=12, family=2, type=1, proto=0, laddr=('0.0.0.0', 0)>
2023-05-09T21:12:57.6724595Z ResourceWarning: Enable tracemalloc to get the object allocation traceback
2023-05-09T21:12:57.6728675Z 2023-05-09 21:12:57,672 |    33 |                 gunicorn.error |    ERROR | Connection in use: ('0.0.0.0', 8060)
2023-05-09T21:12:57.6734120Z 2023-05-09 21:12:57,672 |    33 |                 gunicorn.error |    ERROR | Retrying in 1 second.
2023-05-09T21:12:58.6740294Z 2023-05-09 21:12:58,673 |    33 |                 gunicorn.error |    ERROR | Connection in use: ('0.0.0.0', 8060)
2023-05-09T21:12:58.6740788Z 2023-05-09 21:12:58,673 |    33 |                 gunicorn.error |    ERROR | Retrying in 1 second.
2023-05-09T21:12:59.6747465Z 2023-05-09 21:12:59,674 |    33 |                 gunicorn.error |    ERROR | Connection in use: ('0.0.0.0', 8060)
2023-05-09T21:12:59.6748104Z 2023-05-09 21:12:59,674 |    33 |                 gunicorn.error |    ERROR | Retrying in 1 second.
2023-05-09T21:13:00.6754516Z 2023-05-09 21:13:00,674 |    33 |                 gunicorn.error |    ERROR | Connection in use: ('0.0.0.0', 8060)
2023-05-09T21:13:00.6755618Z 2023-05-09 21:13:00,675 |    33 |                 gunicorn.error |    ERROR | Retrying in 1 second.
2023-05-09T21:13:01.6759062Z 2023-05-09 21:13:01,675 |    33 |                 gunicorn.error |    ERROR | Can't connect to ('0.0.0.0', 8060)
2023-05-09T21:13:25.6810318Z E..............................................
2023-05-09T21:13:25.6810671Z ======================================================================
2023-05-09T21:13:25.6811289Z ERROR: test_memory_limit_separate_per_process (tests.test_integration.IntegrationTests.test_memory_limit_separate_per_process)
2023-05-09T21:13:25.6855942Z Each NsJail process should have its own memory limit.
2023-05-09T21:13:25.6856539Z ----------------------------------------------------------------------
2023-05-09T21:13:25.6856830Z Traceback (most recent call last):
2023-05-09T21:13:25.6857166Z   File "/snekbox/tests/test_integration.py", line 42, in test_memory_limit_separate_per_process
2023-05-09T21:13:25.6857468Z     with run_gunicorn():
2023-05-09T21:13:25.6857739Z   File "/usr/local/lib/python3.11/contextlib.py", line 137, in __enter__
2023-05-09T21:13:25.6858004Z     return next(self.gen)
2023-05-09T21:13:25.6858202Z            ^^^^^^^^^^^^^^
2023-05-09T21:13:25.6858709Z   File "/snekbox/tests/gunicorn_utils.py", line 76, in run_gunicorn
2023-05-09T21:13:25.6859071Z     raise RuntimeError(f"Gunicorn terminated unexpectedly with code {proc.exitcode}.")
2023-05-09T21:13:25.6859436Z RuntimeError: Gunicorn terminated unexpectedly with code 1.
2023-05-09T21:13:25.6859625Z 
2023-05-09T21:13:25.6859862Z ----------------------------------------------------------------------
2023-05-09T21:13:25.6860319Z Ran 67 tests in 29.605s
2023-05-09T21:13:25.6860451Z 
2023-05-09T21:13:25.6860532Z FAILED (errors=1)
2023-05-09T21:13:25.6868264Z 
2023-05-10T03:12:51.7214357Z ##[error]The operation was canceled.
```

--- a/tests/gunicorn_utils.py
+++ b/tests/gunicorn_utils.py
@@ -1,6 +1,7 @@
 import concurrent.futures
 import contextlib
 import multiprocessing
+import time
 from typing import Iterator
 
 from gunicorn.app.wsgiapp import WSGIApplication
@@ -83,4 +84,14 @@ def run_gunicorn(config_path: str = "config/gunicorn.conf.py", **kwargs) -> Iter
 
         yield
     finally:
+        # See https://github.com/python-discord/snekbox/issues/177
+        # Sleeping before terminating the process avoids a case where
+        # terminating the process can take >30 seconds.
+        time.sleep(0.2)
+
         proc.terminate()
+
+        # Actually wait for the process to finish. There doesn't seem to be a
+        # reliable way of checking if the timeout was reached or this ended normally,
+        # but if the timeout is reached it will probably error later anyway.
+        proc.join(timeout=10)

--- a/tests/gunicorn_utils.py
+++ b/tests/gunicorn_utils.py
@@ -92,6 +92,7 @@ def run_gunicorn(config_path: str = "config/gunicorn.conf.py", **kwargs) -> Iter
         proc.terminate()
 
         # Actually wait for the process to finish. There doesn't seem to be a
-        # reliable way of checking if the timeout was reached or this ended normally,
-        # but if the timeout is reached it will probably error later anyway.
+        # reliable way of checking if process ended or the timeout was reached,
+        # so kill the process afterwards to be sure.
         proc.join(timeout=10)
+        proc.kill()

--- a/tests/gunicorn_utils.py
+++ b/tests/gunicorn_utils.py
@@ -33,6 +33,12 @@ def _proc_target(config_path: str, event: multiprocessing.Event, **kwargs) -> No
     def when_ready(_):
         event.set()
 
+    # Clear sys.argv to prevent Gunicorn from trying to interpret the command arguments
+    # used to run the test as it's own arguments.
+    import sys
+
+    sys.argv = [""]
+
     app = _StandaloneApplication(config_path, when_ready=when_ready, **kwargs)
 
     import logging

